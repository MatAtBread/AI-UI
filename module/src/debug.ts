// @ts-ignore
export const DEBUG = globalThis.DEBUG == '*' || globalThis.DEBUG == true || globalThis.DEBUG?.match(/(^|\W)AI-UI(\W|$)/) || false;
export function log(...args:any) {
    if (DEBUG) console.log('(AI-UI)', ...args)
}
