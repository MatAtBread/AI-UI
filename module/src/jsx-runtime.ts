import { ChildTags, TagCreator, tag } from "./ai-ui.js"

/* Support for React.createElement */
const tagCreators = tag();

type AIUIJSXElement<
  N extends (keyof HTMLElementTagNameMap | typeof AIUIJSX | TagCreator<any,any>),
  C extends ChildTags[]
> = N extends typeof AIUIJSX ? C
: N extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[N]
: N extends TagCreator<any,any> ? ReturnType<N>
: never;

function AIUIJSX<
  T extends {},
  N extends (keyof HTMLElementTagNameMap | typeof AIUIJSX | TagCreator<any,any>),
  C extends ChildTags[]
>(tagName: N, attrs: T | null, ...children: C): AIUIJSXElement<N,C> {
  return (tagName === AIUIJSX ? children
    : typeof tagName === 'string'
      ? (tagCreators[tagName as keyof HTMLElementTagNameMap] as any)?.(attrs, ...children)
      : (tagName as TagCreator<any,any>)(attrs, ...children))
}

/* Support for React 17's _jsx(tag,attrs) *
function sterilise<T extends { children?: any }>(attrs: T): Omit<T, 'children'> {
  const childless = { ...attrs };
  delete childless.children;
  return childless;
}

export const jsx = <T extends { children?: any }>(tagName: keyof HTMLElementTagNameMap | Function, attrs: T) =>
  tagName === jsx
    ? attrs.children
    : (typeof tagName === 'string' ? tag([tagName])[tagName] : tagName)(sterilise(attrs), attrs.children)

export const jsxs = jsx;
export const Fragment = jsx;
*/

declare global {
  namespace JSX {
    type Element = AIUIJSXElement<any,any>;
    type IntrinsicElements = {
      [K in keyof HTMLElementTagNameMap]: Partial<HTMLElementTagNameMap[K]>
    }
  }
}

export default {
  AIUIJSX
};
