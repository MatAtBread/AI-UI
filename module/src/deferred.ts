import { DEBUG, console } from "./debug.js";

// Create a deferred Promise, which can be asynchronously/externally resolved or rejected.
export type DeferredPromise<T> = Promise<T> & {
  resolve: (value: T | PromiseLike<T>) => void;
  reject: (value: any) => void;
}

// Used to suppress TS error about use before initialisation
const nothing = (v: any)=>{};

export function deferred<T>(): DeferredPromise<T> {
  let resolve: (value: T | PromiseLike<T>) => void = nothing;
  let reject: (value: any) => void = nothing;
  const promise = new Promise<T>((...r) => [resolve, reject] = r) as DeferredPromise<T>;
  promise.resolve = resolve;
  promise.reject = reject;
  if (DEBUG) {
    const initLocation = new Error().stack;
    promise.catch(ex => (ex instanceof Error || ex?.value instanceof Error) ? console.log("Deferred rejection", ex, "allocated at ", initLocation) : undefined);
  }
  return promise;
}

export function isPromiseLike<T>(x: any): x is PromiseLike<T> {
  return x !== null && x !== undefined && typeof x.then === 'function';
}
