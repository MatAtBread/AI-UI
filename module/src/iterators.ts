import { isAsyncIterable } from "./ai-ui.js";
import { DeferredPromise, deferred } from "./deferred.js";

/* Things to suppliement the JS base AsyncIterable */

export type PushIterator<T> = AsyncExtraIterable<T> & {
  push(value: T): boolean;  // Push a new value to consumers
  close(ex?: Error): void;  // Tell the consumer we're done, with an optional error
};
export type BroadcastIterator<T> = PushIterator<T>;

export interface AsyncExtraIterable<T> extends AsyncIterable<T>, AsyncIterableHelpers {}

/* A function that wraps a "prototypical" AsyncIterator helper, that has `this:AsyncIterable<T>` and returns
  something that's derived from AsyncIterable<R>, result in a wrapped function that accepts
  the same arguments returns a AsyncExtraIterable<X>
*/
function wrapAsyncHelper<T, Args extends any[], R, X extends AsyncIterable<R>>(fn: (this: AsyncIterable<T>, ...args: Args) => X) {
  return function(this: AsyncIterable<T>, ...args:Args) { return withHelpers(fn.call(this, ...args)) }
}

type AsyncIterableHelpers = typeof asyncExtras;
export const asyncExtras = {
  map: wrapAsyncHelper(map),
  filter: wrapAsyncHelper(filter),
  throttle: wrapAsyncHelper(throttle),
  debounce: wrapAsyncHelper(debounce),
  waitFor: wrapAsyncHelper(waitFor),
  count: wrapAsyncHelper(count),
  retain: wrapAsyncHelper(retain),
  broadcast: wrapAsyncHelper(broadcast),
  initially: wrapAsyncHelper(initially),
  consume
};

class QueueIteratableIterator<T> implements AsyncIterableIterator<T> {
  private _pending = [] as DeferredPromise<IteratorYieldResult<T>>[];
  private _items = [] as T[];

  constructor(private readonly stop = () => { }) {
  }

  [Symbol.asyncIterator](this: AsyncIterableIterator<T>) {
    return this;
  }

  next() {
    if (this._items.length) {
      return Promise.resolve({ done: false, value: this._items.shift()! });
    } 
    
    const value = deferred<IteratorYieldResult<T>>();
    this._pending.push(value);
    return value;
  }

  return() {
    const value = { done: true as const, value: undefined };
    if (this._pending) {
      try { this.stop() } catch (ex) { }
      while(this._pending.length)
        this._pending.shift()!.reject(value);
      this._items.splice(0, this._items.length);
    }
    return Promise.resolve(value);
  }

  throw(...args: any[]) {
    const value = { done: true as const, value: args[0] };
    if (this._pending) {
      try { this.stop() } catch (ex) { }
      while(this._pending.length)
        this._pending.shift()!.reject(value);
      this._items.splice(0, this._items.length);
    }
    return Promise.resolve(value);
  }

  push(value: T) {
    if (!this._pending) {
      //throw new Error("pushIterator has stopped");
      return true;
    }
    if (this._pending.length) {
      this._pending.shift()!.resolve({ done: false, value });
    } else {
      this._items.push(value);
    }
    return false;
  }
}

/* An AsyncIterable which typed objects can be published to.
  The queue can be read by multiple consumers, who will each receive
  unique values from the queue (ie: the queue is SHARED not duplicated)
*/
export function pushIterator<T>(stop = () => { }, bufferWhenNoConsumers = false): PushIterator<T> {
  let consumers = 0;
  let ai: QueueIteratableIterator<T> = new QueueIteratableIterator<T>(() => {
    consumers -= 1;
    if (consumers === 0 && !bufferWhenNoConsumers) {
      try { stop() } catch (ex) { }
      // This should never be referenced again, but if it is, it will throw
      (ai as any) = null;
    }
  });

  return Object.assign(Object.create(asyncExtras) as AsyncExtraIterable<T>, {
    [Symbol.asyncIterator]() {
      consumers += 1;
      return ai;
    },
    push(value: T) {
      if (!bufferWhenNoConsumers && consumers === 0) {
        // No one ready to read the results
        return false;
      }
      return ai.push(value);
    },
    close(ex?: Error) {
      ex ? ai.throw?.(ex) : ai.return?.();
      // This should never be referenced again, but if it is, it will throw
      (ai as any) = null;
    }
  });
}

/* An AsyncIterable which typed objects can be published to.
  The queue can be read by multiple consumers, who will each receive
  a copy of the values from the queue (ie: the queue is BROADCAST not shared).

  The iterators stops running when the number of consumers decreases to zero
*/

export function broadcastIterator<T>(stop = () => { }): BroadcastIterator<T> {
  let ai = new Set<QueueIteratableIterator<T>>();

  return <BroadcastIterator<T>>Object.assign(Object.create(asyncExtras) as AsyncExtraIterable<T>, {
    [Symbol.asyncIterator](): AsyncIterableIterator<T> {
      const added = new QueueIteratableIterator<T>(() => {
        ai.delete(added);
        if (ai.size === 0) {
          try { stop() } catch (ex) { }
          // This should never be referenced again, but if it is, it will throw
          (ai as any) = null;
        }
      });
      ai.add(added);
      return withHelpers(added);
    },
    push(value: T) {
      if (!ai?.size)
        return false;

      for (const q of ai.values()) {
        q.push(value);
      }
    },
    close(ex?: Error) {
      for (const q of ai.values()) {
        ex ? q.throw?.(ex) : q.return?.();
      }
      // This should never be referenced again, but if it is, it will throw
      (ai as any) = null;
    }
  });
}

/* Merge asyncIterables into a single asyncIterable */

/* TS hack to expose the return AsyncGenerator a generator of the union of the merged types */
type CollapseIterableType<T> = T[] extends AsyncIterable<infer U>[] ? U : never;
type CollapseIterableTypes<T> = AsyncIterable<CollapseIterableType<T>>;

export const merge = <A extends AsyncIterable<any>[]>(...ai: A) => {
  const it = ai.map(i => Symbol.asyncIterator in i ? i[Symbol.asyncIterator]() : i);
  const promises = it.map((i, idx) => i.next().then(result => ({ idx, result })));
  const results: CollapseIterableType<A[number]>[] = [];
  const forever = new Promise<any>(() => { });
  let count = promises.length;
  const merged: AsyncIterableIterator<A[number]> = {
    [Symbol.asyncIterator]() { return this },
    next() {
      return count 
        ? Promise.race(promises).then(({ idx, result }) => {
          if (result.done) {
            count--;
            promises[idx] = forever;
            results[idx] = result.value;
            return { done: count === 0, value: result.value }
          } else {
            promises[idx] = it[idx].next().then(result => ({ idx, result }));
            return result;
          }
        }).catch(ex => {
          return this.throw?.(ex) ?? Promise.reject({ done: true as const, value: new Error("Iterator merge exception") });
        })
      : Promise.reject({ done: true as const, value: new Error("Iterator merge complete") });
    },
    return() {
      const ex = new Error("Merge terminated");
      for (let i = 0; i < it.length; i++) {
        if (promises[i] !== forever) {
          promises[i] = forever;
          it[i].return?.({ done: true, value: ex }); // Terminate the sources with the appropriate cause
        }
      }
      return Promise.resolve({ done: true, value: ex });
    },
    throw(ex: any) {
      for (let i = 0; i < it.length; i++) {
        if (promises[i] !== forever) {
          promises[i] = forever;
          it[i].throw?.(ex); // Terminate the sources with the appropriate cause
        }
      }
      return Promise.reject(ex);
    }
  };
  return withHelpers(merged as unknown as CollapseIterableTypes<A[number]>);
}


/* 
  Extensions to the AsyncIterable:
  calling `bind(ai)` adds "standard" methods to the specified AsyncIterable
*/

function isExtraIterable<T>(i: any): i is AsyncExtraIterable<T> {
  return isAsyncIterable(i) 
    && Object.keys(asyncExtras)
    .every(k => (k in i) && (i as any)[k] === asyncExtras[k as keyof AsyncIterableHelpers]);
}

// Attach the pre-defined helpers onto an AsyncIterable and return the modified object correctly typed
export function withHelpers<A extends AsyncIterable<any>>(ai: A) {
  if (!isExtraIterable(ai)) {
    Object.assign(ai, asyncExtras);
  }
  return ai as A & (A extends AsyncIterable<infer T> ? AsyncExtraIterable<T> : never)
}

/* AsyncIterable helpers, which can be attached to an AsyncIterator with `withHelpers(ai)`, and invoked directly for foreign asyncIterators */
async function* map<U, R>(this: AsyncIterable<U>, mapper: (o: U) => R | PromiseLike<R>): AsyncIterable<Awaited<R>> {
  const ai = this[Symbol.asyncIterator]();
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(mapper(p.value));
      }
      yield mapper(p.value);
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function* filter<U>(this: AsyncIterable<U>, fn: (o: U) => boolean | PromiseLike<boolean>): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(p.value);
      }
      if (await fn(p.value)) {
        yield p.value;
      }
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function *initially<U, I>(this: AsyncIterable<U>, initValue: I): AsyncIterable<U | I> {
  yield initValue;
  for await (const u of this)
    yield u;
}


async function* throttle<U>(this: AsyncIterable<U>, milliseconds: number): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  let paused = 0;
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(p.value);
      }
      const now = Date.now();
      if (paused < now) {
        paused = now + milliseconds;
        yield p.value;
      }
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

const forever = new Promise<any>(() => { });

// NB: DEBOUNCE IS CURRENTLY BROKEN
async function* debounce<U>(this: AsyncIterable<U>, milliseconds: number): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  let timer: Promise<{ debounced: number, value: U }> = forever;
  let last = -1;
  try {
    while (true) {
      const p = await Promise.race([ai.next(), timer]);
      if ('done' in p && p.done)
        return ai.return?.(p.value);
      if ('debounced' in p && p.debounced) {
        if (p.debounced === last)
          yield p.value;
      } else {
        // We have a new value from the src
        clearTimeout(last);
        timer = new Promise(resolve => {
          last = setTimeout(() => {
            resolve({ debounced: last, value: p.value })
          }, milliseconds);
        });
      }
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function* waitFor<U>(this: AsyncIterable<U>, cb: (done: (...a: unknown[]) => void) => void): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(p.value);
      }
      await new Promise(resolve => cb(resolve));
      yield p.value;
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function* count<U extends {}, K extends string>(this: AsyncIterable<U>, field: K) {
  const ai = this[Symbol.asyncIterator]();
  let count = 0;
  try {
    for await (const value of this) {
      const counted = {
        ...value,
        [field]: count++
      }
      yield counted;
    }
  } catch (ex) {
    throw ex;
  }
}

function retain<U extends {}>(this: AsyncIterable<U>): AsyncIterableIterator<U> & { value: U, done: boolean } {
  const ai = this[Symbol.asyncIterator]();
  let prev: IteratorResult<U, any>;
  return {
    [Symbol.asyncIterator]() { return this },
    next() {
      const n = ai.next();
      n.then(p => prev = p);
      return n;
    },
    return(value?: any) {
      return ai.return?.(value) ?? Promise.resolve({done: true as const, value });
    },
    throw(...args: any[]) {
      return ai.throw?.(args) ?? Promise.resolve({done: true as const, value: args[0] })
    },
    get value() {
      return prev.value;
    },
    get done() {
      return Boolean(prev.done);
    }
  }
}

function broadcast<U, X>(this: AsyncIterable<U>, pipe: ((dest: AsyncIterable<U>) => AsyncIterable<X>) = (x => x as unknown as AsyncIterable<X>)): AsyncIterable<X> {
  const ai = this[Symbol.asyncIterator]();
  const b = broadcastIterator<U>(/*() => console.log("..stooped")*/);
  (function update() {
    ai.next().then(v => {
      if (v.done) {
        // Meh - we throw these away for now.
        // console.log(".broadcast done");
      } else {
        b.push(v.value);
        update();
      }
    }).catch(ex => b.close(ex));
  })();

  return {
    [Symbol.asyncIterator]() {
      const dest = pipe(b);
      return dest[Symbol.asyncIterator]();
    }
  }
}

async function consume<U>(this: AsyncIterable<U>, f: (u: U) => void): Promise<void> {
  for await (const u of this)
    f(u);
}

