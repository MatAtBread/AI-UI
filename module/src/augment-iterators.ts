import { augmentGlobalAsyncGenerators } from './ai-ui.js';
import { AsyncExtraIterable } from './iterators.js';
export * from './iterators.js';

// Find the prototype that holds the [Symbol.asyncIterator] for an async generator function, and add the helpers to it.
augmentGlobalAsyncGenerators();

export {};

declare global {
  interface AsyncGenerator<T = unknown, TReturn = any, TNext = unknown> extends AsyncIterator<T, TReturn, TNext>, AsyncExtraIterable<T> {
  }
}