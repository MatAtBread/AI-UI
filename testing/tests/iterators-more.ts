import { Iterators } from '../../module/src/ai-ui';

const { iterableHelpers } = Iterators;
const sleep = Test.sleep;

let running = true;
async function* g(label: string, t: number, ret = -1, ex = -1) {
  var i = 0;
  try {
    while(running) {
      i += 1;
      console.log(label, "yield", i);
      const y = [label, i] as [string, number];
      yield y;
      if (i===2)
        yield y;
      await sleep(t);
      if (i === ret)
        return label + " return "+i;
      if (i === ex)
        throw label + " throw "+i;
    }
  }
  catch (ex) {
    console.log(label, "Exception", ex);
  }
  finally {
    console.log(label, "finally");
  }
  return label + " done "+i;
}

async function run(i: AsyncIterable<any>, ex = -1) {
  console.log("run:");
  let count = 0;
  try {
    for await (const v of i) {
      if (++count === ex)
        throw "run Exit "+count;
      console.log("next",v);
    }
  } catch (ex) {
    console.log("run ex",ex);
  } finally {
    console.log("run finally");
  }
}

const m = iterableHelpers(g("multi", 0.1, 5, 3)).waitFor(done => setTimeout(done,234)).multi();
const m1 = run(m.map(v => ['first', ...v]));
await sleep(0.1)
const m2 = run(m.map(v => ['second', ...v]));
await Promise.all([m1,m2]);

const n = iterableHelpers(g("consume", 0.1, 5, 3)).waitFor(done => setTimeout(done,234)).multi();
const c1 = n.consume(v => console.log("consume 1st",v));
await sleep(0.1)
const c2 = n.consume(v => console.log("consume 2nd",v));
await Promise.all([c1,c2]);

await run(iterableHelpers(g("waitFor", 0.1, 5)).waitFor(done => setTimeout(done,234)));

await run(iterableHelpers(g("slow", 0.1, 5)).filter(v => sleep(0.5,true)));
await run(g("gen", 0.1, 5));
await run(iterableHelpers(g("filter", 0.1, 5)).filter(v => Boolean(v[1]&1)));
await run(iterableHelpers(g("unique", 0.1, 5)).unique());
await run(iterableHelpers(g("initially", 0.1, 5)).initially(["InitialValue",0]));
await run(iterableHelpers(g("map", 0.1, 5)).map(v => ["mapped",...v]));

await run(g("genThrow", 0.1, 0, 5));
await run(iterableHelpers(g("mapThrow", 0.1, 0, 5)).map(v => ["mapped",...v]));

await run(g("genThrow", 0.1, 5) ,4);
await run(iterableHelpers(g("mapThrow", 0.1, 5)).map(v => ["mapped",...v]), 4);

await run(iterableHelpers(g("map", 0.1, 0, 5)).map(v => ["mapped",...v]));
await run(iterableHelpers(g("mapperEx", 0.1, 9)).map(v => {
  if (v[1] === 3) {
    throw "MapperEX";
  } 
  return ["mapped",...v]})
).then(
  r => console.log("mapperEx then:", r), 
  r => console.log("mapperEx catch:", r));

running = false;
await sleep(2)