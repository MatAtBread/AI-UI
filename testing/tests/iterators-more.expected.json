[
  ["\"run:\""],
  ["\"multi\"","\"yield\"","1"],
  ["\"run:\""],
  ["\"next\"","[\"first\",\"multi\",1]"],
  ["\"next\"","[\"second\",\"multi\",1]"],
  ["\"multi\"","\"yield\"","2"],
  ["\"next\"","[\"first\",\"multi\",2]"],
  ["\"next\"","[\"second\",\"multi\",2]"],
  ["\"next\"","[\"first\",\"multi\",2]"],
  ["\"next\"","[\"second\",\"multi\",2]"],
  ["\"multi\"","\"yield\"","3"],
  ["\"next\"","[\"first\",\"multi\",3]"],
  ["\"next\"","[\"second\",\"multi\",3]"],
  ["\"multi\"","\"Exception\"","\"multi throw 3\""],
  ["\"multi\"","\"finally\""],
  ["\"run finally\""],
  ["\"run finally\""],
  ["\"consume\"","\"yield\"","1"],
  ["\"consume 1st\"","[\"consume\",1]"],
  ["\"consume 2nd\"","[\"consume\",1]"],
  ["\"consume\"","\"yield\"","2"],
  ["\"consume 1st\"","[\"consume\",2]"],
  ["\"consume 2nd\"","[\"consume\",2]"],
  ["\"consume 1st\"","[\"consume\",2]"],
  ["\"consume 2nd\"","[\"consume\",2]"],
  ["\"consume\"","\"yield\"","3"],
  ["\"consume 1st\"","[\"consume\",3]"],
  ["\"consume 2nd\"","[\"consume\",3]"],
  ["\"consume\"","\"Exception\"","\"consume throw 3\""],
  ["\"consume\"","\"finally\""],
  ["\"run:\""],
  ["\"waitFor\"","\"yield\"","1"],
  ["\"next\"","[\"waitFor\",1]"],
  ["\"waitFor\"","\"yield\"","2"],
  ["\"next\"","[\"waitFor\",2]"],
  ["\"next\"","[\"waitFor\",2]"],
  ["\"waitFor\"","\"yield\"","3"],
  ["\"next\"","[\"waitFor\",3]"],
  ["\"waitFor\"","\"yield\"","4"],
  ["\"next\"","[\"waitFor\",4]"],
  ["\"waitFor\"","\"yield\"","5"],
  ["\"next\"","[\"waitFor\",5]"],
  ["\"waitFor\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"slow\"","\"yield\"","1"],
  ["\"next\"","[\"slow\",1]"],
  ["\"slow\"","\"yield\"","2"],
  ["\"next\"","[\"slow\",2]"],
  ["\"next\"","[\"slow\",2]"],
  ["\"slow\"","\"yield\"","3"],
  ["\"next\"","[\"slow\",3]"],
  ["\"slow\"","\"yield\"","4"],
  ["\"next\"","[\"slow\",4]"],
  ["\"slow\"","\"yield\"","5"],
  ["\"next\"","[\"slow\",5]"],
  ["\"slow\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"gen\"","\"yield\"","1"],
  ["\"next\"","[\"gen\",1]"],
  ["\"gen\"","\"yield\"","2"],
  ["\"next\"","[\"gen\",2]"],
  ["\"next\"","[\"gen\",2]"],
  ["\"gen\"","\"yield\"","3"],
  ["\"next\"","[\"gen\",3]"],
  ["\"gen\"","\"yield\"","4"],
  ["\"next\"","[\"gen\",4]"],
  ["\"gen\"","\"yield\"","5"],
  ["\"next\"","[\"gen\",5]"],
  ["\"gen\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"filter\"","\"yield\"","1"],
  ["\"next\"","[\"filter\",1]"],
  ["\"filter\"","\"yield\"","2"],
  ["\"filter\"","\"yield\"","3"],
  ["\"next\"","[\"filter\",3]"],
  ["\"filter\"","\"yield\"","4"],
  ["\"filter\"","\"yield\"","5"],
  ["\"next\"","[\"filter\",5]"],
  ["\"filter\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"unique\"","\"yield\"","1"],
  ["\"next\"","[\"unique\",1]"],
  ["\"unique\"","\"yield\"","2"],
  ["\"next\"","[\"unique\",2]"],
  ["\"unique\"","\"yield\"","3"],
  ["\"next\"","[\"unique\",3]"],
  ["\"unique\"","\"yield\"","4"],
  ["\"next\"","[\"unique\",4]"],
  ["\"unique\"","\"yield\"","5"],
  ["\"next\"","[\"unique\",5]"],
  ["\"unique\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"next\"","[\"InitialValue\",0]"],
  ["\"initially\"","\"yield\"","1"],
  ["\"next\"","[\"initially\",1]"],
  ["\"initially\"","\"yield\"","2"],
  ["\"next\"","[\"initially\",2]"],
  ["\"next\"","[\"initially\",2]"],
  ["\"initially\"","\"yield\"","3"],
  ["\"next\"","[\"initially\",3]"],
  ["\"initially\"","\"yield\"","4"],
  ["\"next\"","[\"initially\",4]"],
  ["\"initially\"","\"yield\"","5"],
  ["\"next\"","[\"initially\",5]"],
  ["\"initially\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"map\"","\"yield\"","1"],
  ["\"next\"","[\"mapped\",\"map\",1]"],
  ["\"map\"","\"yield\"","2"],
  ["\"next\"","[\"mapped\",\"map\",2]"],
  ["\"next\"","[\"mapped\",\"map\",2]"],
  ["\"map\"","\"yield\"","3"],
  ["\"next\"","[\"mapped\",\"map\",3]"],
  ["\"map\"","\"yield\"","4"],
  ["\"next\"","[\"mapped\",\"map\",4]"],
  ["\"map\"","\"yield\"","5"],
  ["\"next\"","[\"mapped\",\"map\",5]"],
  ["\"map\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"genThrow\"","\"yield\"","1"],
  ["\"next\"","[\"genThrow\",1]"],
  ["\"genThrow\"","\"yield\"","2"],
  ["\"next\"","[\"genThrow\",2]"],
  ["\"next\"","[\"genThrow\",2]"],
  ["\"genThrow\"","\"yield\"","3"],
  ["\"next\"","[\"genThrow\",3]"],
  ["\"genThrow\"","\"yield\"","4"],
  ["\"next\"","[\"genThrow\",4]"],
  ["\"genThrow\"","\"yield\"","5"],
  ["\"next\"","[\"genThrow\",5]"],
  ["\"genThrow\"","\"Exception\"","\"genThrow throw 5\""],
  ["\"genThrow\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"mapThrow\"","\"yield\"","1"],
  ["\"next\"","[\"mapped\",\"mapThrow\",1]"],
  ["\"mapThrow\"","\"yield\"","2"],
  ["\"next\"","[\"mapped\",\"mapThrow\",2]"],
  ["\"next\"","[\"mapped\",\"mapThrow\",2]"],
  ["\"mapThrow\"","\"yield\"","3"],
  ["\"next\"","[\"mapped\",\"mapThrow\",3]"],
  ["\"mapThrow\"","\"yield\"","4"],
  ["\"next\"","[\"mapped\",\"mapThrow\",4]"],
  ["\"mapThrow\"","\"yield\"","5"],
  ["\"next\"","[\"mapped\",\"mapThrow\",5]"],
  ["\"mapThrow\"","\"Exception\"","\"mapThrow throw 5\""],
  ["\"mapThrow\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"genThrow\"","\"yield\"","1"],
  ["\"next\"","[\"genThrow\",1]"],
  ["\"genThrow\"","\"yield\"","2"],
  ["\"next\"","[\"genThrow\",2]"],
  ["\"next\"","[\"genThrow\",2]"],
  ["\"genThrow\"","\"yield\"","3"],
  ["\"genThrow\"","\"finally\""],
  ["\"run ex\"","\"run Exit 4\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"mapThrow\"","\"yield\"","1"],
  ["\"next\"","[\"mapped\",\"mapThrow\",1]"],
  ["\"mapThrow\"","\"yield\"","2"],
  ["\"next\"","[\"mapped\",\"mapThrow\",2]"],
  ["\"next\"","[\"mapped\",\"mapThrow\",2]"],
  ["\"mapThrow\"","\"yield\"","3"],
  ["\"mapThrow\"","\"finally\""],
  ["\"run ex\"","\"run Exit 4\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"map\"","\"yield\"","1"],
  ["\"next\"","[\"mapped\",\"map\",1]"],
  ["\"map\"","\"yield\"","2"],
  ["\"next\"","[\"mapped\",\"map\",2]"],
  ["\"next\"","[\"mapped\",\"map\",2]"],
  ["\"map\"","\"yield\"","3"],
  ["\"next\"","[\"mapped\",\"map\",3]"],
  ["\"map\"","\"yield\"","4"],
  ["\"next\"","[\"mapped\",\"map\",4]"],
  ["\"map\"","\"yield\"","5"],
  ["\"next\"","[\"mapped\",\"map\",5]"],
  ["\"map\"","\"Exception\"","\"map throw 5\""],
  ["\"map\"","\"finally\""],
  ["\"run finally\""],
  ["\"run:\""],
  ["\"mapperEx\"","\"yield\"","1"],
  ["\"next\"","[\"mapped\",\"mapperEx\",1]"],
  ["\"mapperEx\"","\"yield\"","2"],
  ["\"next\"","[\"mapped\",\"mapperEx\",2]"],
  ["\"next\"","[\"mapped\",\"mapperEx\",2]"],
  ["\"mapperEx\"","\"yield\"","3"],
  ["\"mapperEx\"","\"Exception\"","\"MapperEX\""],
  ["\"mapperEx\"","\"finally\""],
  ["\"run ex\"","{\"done\":true,\"value\":\"MapperEX\"}"],
  ["\"run finally\""],
  ["\"mapperEx then:\"",null]
]